kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,270 @@
+    local version = lower(v.version)
+    if versions_set[version] then
+      return v
+    end
+  end
+
+  return { name = name, description = "No valid version" }
+end
+
+local function negotiate_service(name, versions)
+  name = lower(name)
+
+  if type(versions) ~= "table" then
+    error("invalid versions array for service " .. name)
+  end
+
+  local supported_service = supported_services[name]
+  if not supported_service then
+    return {
+      description = "unknown service.",
+    }
+  end
+
+  local service_response = negotiate_version(name, versions, supported_service)
+  return service_response
+end
+
+local function log_negotiation_result(name, version)
+  local ok = version.version ~= nil
+  ngx_log(NOTICE, _log_prefix, "service ",
+    (ok and "accepted" or "rejected"),
+    ": \"", name, "\"",
+    (ok and ", version: " .. version.version or ""),
+    ", ", (ok and "description" or "reason"), ": ", version.description
+  )
+end
+
+local function negotiate_services(services_requested)
+  local services = {}
+
+  for name, versions in pairs(services_requested) do
+    if type(name) ~= "string" or type(versions) ~= "table" then
+      error("malformed service requested " .. name)
+    end
+
+    local negotiated_version = negotiate_service(name, versions)
+    services[name] = negotiated_version
+
+    log_negotiation_result(name, negotiated_version)
+  end
+
+  return services
+end
+
+
+local function register_client(cluster_data_plane_purge_delay, id, client_node)
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = id, }, {
+    last_seen = ngx.time(),
+    config_hash = DECLARATIVE_EMPTY_CONFIG_HASH,
+    hostname = client_node.hostname,
+    ip = ngx.var.remote_addr,
+    version = client_node.version,
+    sync_status = client_node.sync_status,
+  }, { ttl = cluster_data_plane_purge_delay })
+
+  if not ok then
+    ngx_log(ERR, _log_prefix, "unable to update clustering data plane status: ", err)
+    return error(err)
+  end
+end
+
+function _M.init_negotiation_server(service, conf)
+  service:import("kong.services.negotiation.v1.negotiation")
+  service:set_handler("NegotiationService.NegotiateServices", function(peer, nego_req)
+    local ok, result = pcall(function()
+      local dp_id = peer.id
+      ngx_log(NOTICE, "negotiating services for DP: ", dp_id)
+      local services_requested = unwrap_services(nego_req.services_requested)
+      verify_request(nego_req)
+
+      nego_req.node.sync_status = verify_node_compatibility(nego_req.node)
+      local services = negotiate_services(services_requested)
+      register_client(conf.cluster_data_plane_purge_delay, dp_id, nego_req.node)
+
+      local nego_result = {
+        services = services
+      }
+
+      return nego_result
+    end)
+
+    if not ok then
+      ngx_log(ERR, _log_prefix, result)
+      return { error_message = result }
+    end
+
+    return result
+  end)
+end
+
+-- TODO: use event to notify other workers
```

Comment:
or just set in a shared resource (shdict, lmdb)

Reply:
The problem is `shdict` would block all workers.

Reply:
I agree. Let's decide it after this PR.